packages/ses/src/module-static-record.js

```diff
@@ -0,0 +1,44 @@
+import * as babel from '@agoric/babel-standalone';
+// We are using the above import form, and referring to its default export
+// explicitly below like babel.default, because the proper construct causes a
+// Rollup error.
+// This form:
+//   import babel from '@agoric/babel-standalone';
+// And this variant:
+//   import { default as babel } from '@agoric/babel-standalone';
+// Both produce:
+//   Error: 'default' is not exported by .../@agoric/babel-standalone/babel.js
+import { makeModuleAnalyzer } from '@agoric/transform-module';
+
+const analyze = makeModuleAnalyzer(babel.default);
+
+// moduleAnalyses are the private data of a ModuleStaticRecord.
+// We use moduleAnalyses in the loader/linker to look up
+// the analysis corresponding to any ModuleStaticRecord constructed by an
+// importHook.
+// This WeakMap is shared among modules in this package, but not exposed to the
+// public.
+export const moduleAnalyses = new WeakMap();
+
+// ModuleStaticRecord captures the effort of parsing and analyzing module text
+// so a cache of ModuleStaticRecords may be shared by multiple Compartments.
+export class ModuleStaticRecord {
```

Comment:
Given the layering we're going for, it probably makes sense that the exported `ModuleStaticRecord` class is not frozen. Neither is `ModuleStaticRecord.prototype` or anything hung off the class or prototype. That's because we're shimming compartments as we'll propose for EcmaScript in general, outside of SES.

A consequence is that we need to arrange for `lockdown` to harden these.

Reply:
I can tack this onto this change. We have options.

Since this is not a global object, it doesn’t get captured automatically. If it were communicated as a global object, I imagine I’d just have to whitelist it.

So, one option is for the compartment-shim to reveal Compartment and ModuleStaticRecord as globals instead of exporting them.

The other option is to harden them as one-offs in lockdown-shim.js.

Reply:
I'm really conflicted here. Given the current ecma262 spec, the natural way to introduce another user-visible class-like powerless abstraction such as `ModuleStaticRecord` is to make the constructor a shared global. Then it would fit into the story we already have for how `lockdown` interacts with any other shared global.

The alternative is to provide these as shims of proposed built-in modules. If the built-in module instances are to typically be shared across the realm, just as primordials are, then after `lockdown` the shared powerless built-in modules must be pure. We have several options here but haven't yet taken on this part of the design. We've been able to postpone it for the moment, since EcmaScript does not yet have built-in modules.

I think I just talked myself into the first, at least at this stage. `ModuleStaticRecord` should be shimmed as yet another normal class-like powerless global.

The only other emulated built-in module we have as part of SES is `harden`. `harden` is not class-like. Given what `harden` is about, the shared module instance exporting `harden` can effectively be pure.

Reply:
Does the `ModuleStaticRecord` as shimmed global, to be added to the whitelists, seem like a good way to solve the lockdown integration for now? This is the only remaining issue with this PR.

Reply:
I’ve added a change that adds ModuleStaticRecord to the lockdown whitelists in the same way Compartment is there already. I would like to follow-up with a change to pivot SES from exporting its API to dropping it on `globalThis`. https://github.com/Agoric/SES-shim/issues/282

I believe this addresses your feedback. There remains one broken integration test, again due to an interaction between Babel and Rollup.